flink-connectors/flink-connector-pubsub/src/main/java/org/apache/flink/streaming/connectors/pubsub/DefaultPubSubSubscriberFactory.java

```diff
@@ -0,0 +1,67 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.streaming.connectors.pubsub;
+
+import org.apache.flink.streaming.connectors.pubsub.common.PubSubSubscriberFactory;
+
+import com.google.api.gax.batching.FlowControlSettings;
+import com.google.api.gax.batching.FlowController;
+import com.google.api.gax.core.FixedCredentialsProvider;
+import com.google.api.gax.grpc.GrpcTransportChannel;
+import com.google.api.gax.rpc.FixedTransportChannelProvider;
+import com.google.api.gax.rpc.TransportChannel;
+import com.google.auth.Credentials;
+import com.google.cloud.pubsub.v1.MessageReceiver;
+import com.google.cloud.pubsub.v1.Subscriber;
+import com.google.pubsub.v1.ProjectSubscriptionName;
+import io.grpc.ManagedChannel;
+import io.grpc.ManagedChannelBuilder;
+
+final class DefaultPubSubSubscriberFactory implements PubSubSubscriberFactory {
+	private String hostAndPort;
+
+	void withHostAndPort(String hostAndPort) {
+		this.hostAndPort = hostAndPort;
+	}
+
+	@Override
+	public Subscriber getSubscriber(Credentials credentials, ProjectSubscriptionName projectSubscriptionName, MessageReceiver messageReceiver) {
+		FlowControlSettings flowControlSettings = FlowControlSettings.newBuilder()
+			.setMaxOutstandingElementCount(10000L)
+			.setMaxOutstandingRequestBytes(100000L)
```

Comment:
final hostAndPort has been fixed.

Reply:
In my own applications I change/tune 3 parameters:

- setMaxOutstandingElementCount
- setMaxOutstandingRequestBytes
- setMaxAckExtensionPeriod

The first two go hand in hand. They are what you would tune when dealing with backpressure (see the flink docs pubsub section I added for a bit of explanation).
The third indicates how long a message is kept unacknowledged. After this time PubSub might start sending the message to other consumers. This is relevant when the time between checkpoints is high.

My thought was to not expose this through the Source builder because it's kinda tricky to properly explain what they do, while in the context of a `Subscriber` the google docs speak for themselves.

I do see your point though: it would increase usability a lot if people can just tune the number a bit without having to go through the hassle of figuring out how to create a `Subscriber`.

So I now added:

		/**
		 * Tune how many messages the connector will buffer when the Flink pipeline cannot keep up (backpressure).
		 *
		 * @param maxMessagesReceivedNotProcessed This indicates how many messages will be read and buffered until when the flink pipeline can't handle the messages fast enough.
		 * @param maxBytesReceivedNotProcessed This indicates how many bytes will be read and buffered. A good pick would be: maxBytesReceivedNotProcessed = maxMessagesReceivedNotProcessed * averageBytesSizePerMessage
		 */
		public PubSubSourceBuilder<OUT> withBackpressureParameters(long maxMessagesReceivedNotProcessed, long maxBytesReceivedNotProcessed) {
			this.maxMessagesReceivedNotProcessed = maxMessagesReceivedNotProcessed;
			this.maxBytesReceivedNotProcessed = maxBytesReceivedNotProcessed;
			return this;
		}

Do you think this is clear enough?

Reply:
@Xeli, the proposition is to allow the modification of these parameters via the Factory. That is, having the default Factory, which one can tweak. As mentioned below, this also allows to hide the  testing options (hostAndPort) from the main API (builder).

Reply:
Thanks for updating the docs!

I think we should consider moving all `DefaultPubSubSubscriberFactory` specific config parameters out of the `PubSubSourceBuilder`, as @afedulov said.

Reply:
Hmm, so we would have 3 ways of initializing the connector, right?

Using all the default values:

    PubSubSource.newBuilder(schema, project, subscription)
                             .withCredentials(someCredentials())
                             .build();

Using defaultFactory:

    PubSubSubscriberFactory factory = DefaultPubSubSubscriberFactory.newBuilder()                                                                                                                          
        .withHostAndPort("localhost:1234")
        .withMaxOutstandingElementCount(1)
        .maxOutstandingRequestBytes(12)
        .build();
    PubSubSource.newBuilder(schema, project, subscription)
                             .withCredentials(someCredentials())
                             .withPubSubSubscriberFactory(factory)
                             .build();

If someone wants complete control:

    PubSubSubscriberFactory factory = (credentials, projectSubscription, messageReceiver) -> {
        Subscriber.Builder builder = Subscriber
			etc etc;
       return builder.build();
    };
    PubSubSource.newBuilder(schema, project, subscription)
                             .withCredentials(someCredentials())
                             .withPubSubSubscriberFactory(factory)
                             .build();

Which basically means I remove the two builder methods from PubSubSource, add a builder to the default factory, and make default factory public.

Reply:
Yes, I think that would be a better separation.

Reply:
Now that I switched to the sync API I created 2 PubSubSubscriberFactories. The options now are:

Using all the default values:

```
PubSubSource.newBuilder(schema, project, subscription)
                         .withCredentials(someCredentials())
                         .build();
```
Using emulator:

```
PubSubSource.newBuilder(schema, project, subscription)
                         .withCredentials(someCredentials())
	                 .withPubSubSubscriberFactory(new PubSubSubscriberFactoryForEmulator("localhost:1234))
                         .build();
```

If someone wants complete control:

```
PubSubSubscriberFactory factory = (credentials, projectSubscription, messageReceiver) -> {
    Subscriber.Builder builder = Subscriber
		etc etc;
   return builder.build();
};
PubSubSource.newBuilder(schema, project, subscription)
                         .withCredentials(someCredentials())
                         .withPubSubSubscriberFactory(factory)
                         .build();
```